File: MotorcycleContest/MotorcycleContest/Repository/UserDbRepository.cs
using System.Collections.ObjectModel;
using System.Data;
using log4net;
using MotorcycleContest.Domain;

namespace MotorcycleContest.Repository;

public class UserDbRepository : AbstractDbRepository<Guid, User>, IUserRepository
{
    private static readonly ILog Log = LogManager.GetLogger("UserDbRepository");

    public UserDbRepository(IDictionary<string, string> properties) : base(properties)
    {
        Log.Info("Initializing the UserDbRepository...");
    }

    public override User? FindOne(Guid tid)
    {
        return null;
    }

    public override Collection<User>? FindAll()
    {
        return null;
    }

    public override User? Save(User te)
    {
        return null;
    }

    public override User? Delete(User te)
    {
        return null;
    }

    public override User? Update(User te)
    {
        return null;
    }

    public User? FindUserByCredentials(string username, string password)
    {
        Log.InfoFormat("Finding user with username {0} and password {1}", username, password);

        Log.Info("Establishing a connection with the database...");
        using (var connection = DBUtils.GetConnection(Properties))
        {
            Log.Info("Connection established.");
            Log.Info("Creating a command...");
            using (var command = connection.CreateCommand())
            {
                Log.Info("Command created.");
                command.CommandText = "SELECT uid, first_name, last_name, username, password FROM users WHERE username = @username";

                Log.Info("Populating the command with parameters...");
                var usernameParam = command.CreateParameter();
                usernameParam.ParameterName = "@username";
                usernameParam.Value = username;
                command.Parameters.Add(usernameParam);

                Log.Info("Executing the command...");
                using (var dataReader = command.ExecuteReader())
                {
                    Log.Info("Command executed.");

                    if (dataReader.Read())
                    {
                        Log.Info("User found.");
                        Log.Info("Checking the password...");
                        var hashedPassword = dataReader.GetString(4);

                        if (!CheckPassword(password, hashedPassword))
                        {
                            Log.Error("Password is incorrect.");
                            return null;
                        }

                        Log.Info("Password is correct. Initializing the user.");

                        var id = dataReader.GetGuid(0);
                        var firstName = dataReader.GetString(1);
                        var lastName = dataReader.GetString(2);

                        var user = new User(id, firstName, lastName, username);

                        Log.InfoFormat("User initialized. Returning with {0}", user);
                        return user;
                    }
                }
            }
        }

        Log.InfoFormat("No user found with username {0} and password {1}.", username, password);
        return null;
    }


    protected override User ExtractFromDataReader(IDataReader dataReader)
    {
        Log.Info("Extracting user from data reader...");
        var id = dataReader.GetGuid(0);
        var firstName = dataReader.GetString(1);
        var lastName = dataReader.GetString(2);
        var username = dataReader.GetString(3);

        Log.Info("User extracted.");
        return new User(id, firstName, lastName, username);
    }

    private static bool CheckPassword(string password, string hashedPassword)
    {
        Log.InfoFormat("Checking password {0} against hashed password {1}", password, hashedPassword);
        return BCrypt.Net.BCrypt.Verify(password, hashedPassword);
    }
}

File: mpp-proiect-java-culbec/MotorcycleContest/Persistence/src/main/java/Repository/UserDBRepository.java
package Repository;

import ConnectionUtils.DBUtils;
import Model.User;
import org.springframework.security.crypto.bcrypt.BCrypt;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Optional;

public class UserDBRepository extends DBRepository<Integer, User> implements UserRepository {
    public UserDBRepository(DBUtils dbUtils) throws RepositoryException {
        super(dbUtils);
    }

    @Override
    protected void initPreparedStatements() throws RepositoryException {
        logger.traceEntry("Initializing the prepared statements for the Model.User Repository...");

        String sqlSelectOnUsername = "select * from users u where u.username = ?";

        logger.info("Trying to establish a connection with the database...");
        Connection connection = this.dbUtils.getConnection();

        if (connection == null) {
            logger.error("Couldn't establish the connection with the database!");
            throw new RepositoryException("Couldn't establish the connection with the database!");
        }

        logger.info("Inserting the prepared statements...");
        try {
            this.preparedStatementMap.put("selectOnUsername", connection.prepareStatement(sqlSelectOnUsername));
        } catch (SQLException e) {
            logger.error("Couldn't insert the prepared statements: {}", e.getMessage());
            throw new RepositoryException("Couldn't initialize the prepared statements for the Model.User Repository: " + e.getMessage());
        }

        logger.traceExit("Initialized the prepared statements for the Model.User Repository!");
    }

    @Override
    protected User extractFromResultSet(ResultSet resultSet) throws RepositoryException {
        logger.traceEntry("Extracting the user from the result set...");

        try {
            Integer uid = resultSet.getInt("uid");
            String firstName = resultSet.getString("first_name");
            String lastName = resultSet.getString("last_name");
            String username = resultSet.getString("username");

            User user = new User.Builder()
                    .setId(uid)
                    .setFirstName(firstName)
                    .setLastName(lastName)
                    .setUsername(username)
                    .build();
            logger.traceExit("Extracted the user from the result set!");
            return user;
        } catch (SQLException e) {
            logger.error("Couldn't extract the user from the result set: {}", e.getMessage());
            throw new RepositoryException("Couldn't extract the user from the result set: " + e.getMessage());
        }
    }

    @Override
    public Optional<User> findOne(Integer id) {
        return Optional.empty();
    }

    @Override
    public Iterable<User> findAll() {
        return null;
    }

    @Override
    public Optional<User> save(User user) {
        return Optional.empty();
    }

    @Override
    public Optional<User> delete(User user) {
        return Optional.empty();
    }

    @Override
    public Optional<User> update(User user) {
        return Optional.empty();
    }

    @Override
    public Optional<User> findUserByCredentials(String username, String password) throws RepositoryException {
        logger.traceEntry("Trying to find the user with the credentials {} {} ", username, password);

        try {
            PreparedStatement preparedStatement = this.preparedStatementMap.get("selectOnUsername");
            preparedStatement.setString(1, username);

            ResultSet resultSet = preparedStatement.executeQuery();

            if (resultSet.next()) {
                String _password = resultSet.getString("password");

                User user = this.extractFromResultSet(resultSet);

                if (user != null && checkPassword(password, _password)) {
                    logger.traceExit("Found the user with the credentials: " + username + " - " + password);
                    return Optional.of(user);
                }
            }
        } catch (SQLException e) {
            logger.error("Couldn't find the user with the credentials: {} - {} - {}", username, password, e.getMessage());
            throw new RepositoryException("Couldn't find the user with the credentials: " + username + " - " + password + " - " + e.getMessage());
        }

        logger.traceExit("Couldn't find the user with the credentials: " + username + " - " + password);
        return Optional.empty();
    }

    /**
     * Checks if the passed password is the same as the encrypted one.
     *
     * @param password       Password to be checked.
     * @param hashedPassword Encrypted password.
     * @return {@code true} if the password is the same as the encrypted one, {@code false} otherwise.
     */
    private boolean checkPassword(String password, String hashedPassword) {
        logger.traceEntry("Checking the password: " + password + " - " + hashedPassword);
        logger.traceExit("Checked the password: " + password + " - " + hashedPassword);
        return BCrypt.checkpw(password, hashedPassword);
    }

}
